### PR TITLE
Modernize plugin for new UI support

### LINKED ISSUE #7: Both Label and Text no more visible in PyCharm Professional 20.1
Here's what my bottom project window in PyCharm Professional 20.1 after installing this package. And using it in Ubuntu 20.04

image

### LINKED ISSUE #8: In the beta UI of Rider 2022.3 it only shows in one window
It's a beautiful little feature, but for some reason, but if you have multiple solutions open, it only shows up on one screen.

### LINKED ISSUE #9: doesnt work in "new u"i updated in latest webstorm

### PR DESCRIPTION
**Pull request by simschla**
Here is my proposal to modernize the infrastructural part of the plugin (first steps)

- Use jetbrain's plugin-build framework (gradle + kotlin)
- Adapt the code to work with the "new" and the "old" UI stylewise (font sizes!) 
- Add the StatusBar widget using the recommended `WidgetFactory` approach (removes race condition and resolves #9 - possibly resolves #7 and resolves #8 but did not explicitly test that)
- bump to use newer intellij versions as well as Java 17 (runtime jvm of intellij version)

I did not test the signing/publishing part of the gradle build on my end, that's something that you might be able to test @drinchev ?

### CODE REVIEW

**Comment by drinchev** [comment]
Just checked out the code. Looks amazing. Thanks so much for this PR.

I will try to figure out the release part over the weekend.

Do you mind adding you as a contributor in the README?

**Comment by simschla** [comment]
Nice, thanks for merging. Glad I could help.

I'm perfectly fine with being added to the README 👍 

I'd also be down helping you to drive the plugin further, since I use it every day 😄  But of course, only if that's something that you'd be up to.

**Comment by drinchev** [comment]
Just a heads up, regarding the publish status.

Today, I received a review from IntelliJ and they require to get rid of the internal API usage : 

'org.jetbrains.annotations.ApiStatus.Internal'

I will try to update the code, according to the review feedback this week.

**Comment by simschla** [comment]
Thanks for the heads-up. I did not intentionally use an internal API. Please let me know If you need me to look into it.

File: src/main/java/com/drinchev/projectlabel/utils/UtilsFont.java [hunk 1]
```diff
@@ -42,4 +43,13 @@ public static Font setAttributes(Font font, Float weight, Float size) {
         return font.deriveFont(attributes);
     }
 
+    public static JBFont getStatusBarItemFont() {
+        try {
+            return JBFont.smallOrNewUiMedium();
```
Comment (drinchev) [inline src/main/java/com/drinchev/projectlabel/utils/UtilsFont.java:48]: I'm looking at the code and I also have this error. 

Was wondering what bad can happen if we just replace this with `.medium()` and call it a day? :D 

image

Comment (simschla) [reply src/main/java/com/drinchev/projectlabel/utils/UtilsFont.java:48]: The javadoc of `smallOrNewUiMedium()` explains why I used it:

> In new UI small() font should be used only in exceptional cases when the medium size won't fit at all. This is a temporary method for backward compatibility with old UI, which returns medium() for new UI and small() for old UI. Will be replaced by medium() after full migration to new UI

So, if we just use `medium()` we will break the UI for folks that have not migrated to the "new UI".

How about we do something with reflection?

```java
    public static JBFont getStatusBarItemFont() {
        try {
            // while the new UI is experimental, we have to carefully choose the font size
            if (JBFont.class.getDeclaredMethod("smallOrNewUiMedium") != null) {
                return Registry.is("ide.experimental.ui") ? JBFont.medium() : JBFont.small();
            }
        } catch (NoSuchMethodException e) {
            return JBFont.medium(); // the method has been removed, medium should be the new normal
        }
        throw new IllegalStateException("Unable to determine the status bar font size");
    }
```

Comment (drinchev) [reply src/main/java/com/drinchev/projectlabel/utils/UtilsFont.java:48]: (Un)fortunately the plugin verification passed with `JBFont.medium()` before I was able to read your comment.

Reading [an official reply](https://intellij-support.jetbrains.com/hc/en-us/community/posts/8954459501970-How-to-determine-if-new-ui-is-enabled-), it seems that `Registry.is("ide.experimental.ui")` might break in the future. I hope there's a better way of detecting this, however if I don't find anything else I will give this a try :)

**Pull request merged**